### PR TITLE
Enhance the segment span validation

### PR DESCRIPTION
**Pull request by mrproliu**
Currently, in the Go version of the agent, the order of spans in a segment can become unstable due to fast execution, although the span IDs are always ordered. To address this issue, this PR is submitted to allow validation of corresponding data in the segment when spans are out of order based on their span IDs.

### CODE REVIEW

**Review by wu-sheng — APPROVED** [review]

**Comment by kezhenxu94** [comment]
> Currently, in the Go version of the agent, the order of spans in a segment can become unstable due to fast execution, although the span IDs are always ordered. To address this issue, this PR is submitted to allow validation of corresponding data in the segment when spans are out of order based on their span IDs.

If it's the order's matter, can we just sort them out every time before we do an actual validation?

**Comment by mrproliu** [comment]
Yes, We could. I'm just concerned about whether it will affect the existing validation. If you think sorting is enough, I can change it to validate after sorting.

**Comment by kezhenxu94** [comment]
> Yes, We could. I'm just concerned about whether it will affect the existing validation. If you think sorting is enough, I can change it to validate after sorting.

I believe it doesn't affect the existing validation because we have been assuming the spans are in order.

BTW, why don't we switch to the infra-e2e project for validation?

**Comment by mrproliu** [comment]
> I believe it doesn't affect the existing validation because we have been assuming the spans are in order.

Sure, I could do an update for this.

> BTW, why don't we switch to the infra-e2e project for validation?

They have different processes, such as agent tests needing to change framework versions in customized configuration files, and they contain different health checkers, etc.

f we want to switch to infra-e2e, there may still be a lot of customization work that needs to be done, but I will consider how to handle it. :)

**Comment by wu-sheng** [comment]
> BTW, why don't we switch to the infra-e2e project for validation?

Because it doesn't have capabilities to roll the versions(supported versions).

**Comment by mrproliu** [comment]
> I believe it doesn't affect the existing validation because we have been assuming the spans are in order.

BTW, I think maybe we cannot sort it, because the excepted span id may be an expression.

File: validator/src/main/java/org/apache/skywalking/plugin/test/agent/tool/validator/assertor/SegmentAssert.java [hunk 1]
```diff
@@ -90,7 +91,28 @@ private static boolean spansEquals(List<Span> excepted, List<Span> actual) {
...
                 spanEquals(exceptedSpan, actualSpan);
             } catch (AssertFailedException e) {
-                throw new SpanAssertFailedException(e, exceptedSpan, actualSpan);
+                if (Objects.equals(exceptedSpan.spanId(), actualSpan.spanId())) {
+                    throw new SpanAssertFailedException(e, exceptedSpan, actualSpan);
+                }
+                // if the span id is not equals, trying to find the span by span id
+                Span tmpSpan = null;
+                for (Span s : actual) {
+                    if (Objects.equals(exceptedSpan.spanId(), s.spanId())) {
```
Comment (kezhenxu94) [inline validator/src/main/java/org/apache/skywalking/plugin/test/agent/tool/validator/assertor/SegmentAssert.java:100]: Is `exceptedSpan.spanId()` an expression ? @mrproliu

**Review by kezhenxu94 — APPROVED** [review]

**Comment by mrproliu** [comment]
The GHA is always Queued, so I cannot merge this PR. Let me create another PR for trying to run the CI.
image

**Pull request closed without merging**